File: ios/Shared/RelayDescriptor.swift
import Foundation

public enum RelayConstants {
    public static let protocolVLESSRealityVision = "vless-reality-vision"
    public static let flowVision = "xtls-rprx-vision"
    public static let exitModeDirect = "direct"
    public static let transportDirect = "direct"
    public static let nodeClassFoundation = "foundation"
    public static let nodeClassVolunteer = "volunteer"
}

/// One canonical WSS/CDN front from the broker-signed relay descriptor.
public struct WssFrontDescriptor: Codable, Equatable, Sendable {
    public let id: String
    public let url: String
    public let protocolVersion: Int

    public init(id: String, url: String, protocolVersion: Int) {
        self.id = id
        self.url = url
        self.protocolVersion = protocolVersion
    }

    enum CodingKeys: String, CodingKey, CaseIterable {
        case id
        case url
        case protocolVersion = "protocol_version"
    }

    /// Reject nested schema extensions before typed decoding can erase them. Top-level relay
    /// descriptors remain forward-compatible, but every signed front must exactly match wsscore.
    public init(from decoder: Decoder) throws {
        let raw = try decoder.container(keyedBy: AnyCodingKey.self)
        let allowed = Set(CodingKeys.allCases.map(\.rawValue))
        let unknown = raw.allKeys.map(\.stringValue).filter { allowed.contains($0) == false }
        guard unknown.isEmpty else {
            throw DecodingError.dataCorrupted(
                DecodingError.Context(
                    codingPath: decoder.codingPath,
                    debugDescription: "WSS front contains unknown fields: \(unknown.sorted())"
                )
            )
        }
        let values = try decoder.container(keyedBy: CodingKeys.self)
        id = try values.decode(String.self, forKey: .id)
        url = try values.decode(String.self, forKey: .url)
        protocolVersion = try values.decode(Int.self, forKey: .protocolVersion)
    }

    public func encode(to encoder: Encoder) throws {
        var values = encoder.container(keyedBy: CodingKeys.self)
        try values.encode(id, forKey: .id)
        try values.encode(url, forKey: .url)
        try values.encode(protocolVersion, forKey: .protocolVersion)
    }

    private struct AnyCodingKey: CodingKey {
        let stringValue: String
        let intValue: Int?

        init?(stringValue: String) {
            self.stringValue = stringValue
            intValue = nil
        }

        init?(intValue: Int) {
            stringValue = String(intValue)
            self.intValue = intValue
        }
    }
}

public struct RelayDescriptor: Codable, Equatable, Identifiable, Sendable {
    public let id: String
    /// Friendly relay name (operator-supplied or generated); absent on older brokers.
    public let label: String?
    public let publicHost: String
    public let publicPort: Int
    public let relayProtocol: String
    public let clientID: String
    public let realityPublicKey: String
    public let shortID: String
    public let serverName: String
    public let flow: String
    public let exitMode: String
    public let maxSessions: Int
    public let maxMbps: Int
    /// Software version for any relay class; coding key preserves the legacy broker wire field.
    public let relayVersion: String
    /// Broker-attested trust class. Descriptors predating this field are volunteer relays.
    public let nodeClass: String
    /// Empty on legacy descriptors; otherwise "direct" or "tunnel".
    public let transport: String
    /// Exact canonical WSS fronts covered by the relay-list signature.
    public let wssFronts: [WssFrontDescriptor]
    public let registeredAt: Date
    public let lastHeartbeatAt: Date
    public let expiresAt: Date
    /// Broker-served exit location, absent until the broker's geo lookup succeeds (older brokers
    /// never send it). For tunnel (CGNAT) relays this is where traffic actually exits, which is
    /// NOT `publicHost` (the relay hub) — never geolocate `publicHost` client-side.
    public let city: String?
    public let country: String?
    public let countryCode: String?
    public let latitude: Double?
    public let longitude: Double?

    public init(
        id: String,
        label: String? = nil,
        publicHost: String,
        publicPort: Int,
        relayProtocol: String,
        clientID: String,
        realityPublicKey: String,
        shortID: String,
        serverName: String,
        flow: String,
        exitMode: String,
        maxSessions: Int,
        maxMbps: Int,
        relayVersion: String,
        nodeClass: String = RelayConstants.nodeClassVolunteer,
        transport: String = "",
        wssFronts: [WssFrontDescriptor] = [],
        registeredAt: Date,
        lastHeartbeatAt: Date,
        expiresAt: Date,
        city: String? = nil,
        country: String? = nil,
        countryCode: String? = nil,
        latitude: Double? = nil,
        longitude: Double? = nil
    ) {
        self.id = id
        self.label = label
        self.publicHost = publicHost
        self.publicPort = publicPort
        self.relayProtocol = relayProtocol
        self.clientID = clientID
        self.realityPublicKey = realityPublicKey
        self.shortID = shortID
        self.serverName = serverName
        self.flow = flow
        self.exitMode = exitMode
        self.maxSessions = maxSessions
        self.maxMbps = maxMbps
        self.relayVersion = relayVersion
        self.nodeClass = nodeClass
        self.transport = transport
        self.wssFronts = wssFronts
        self.registeredAt = registeredAt
        self.lastHeartbeatAt = lastHeartbeatAt
        self.expiresAt = expiresAt
        self.city = city
        self.country = country
        self.countryCode = countryCode
        self.latitude = latitude
        self.longitude = longitude
    }

    enum CodingKeys: String, CodingKey {
        case id
        case label
        case publicHost = "public_host"
        case publicPort = "public_port"
        case relayProtocol = "protocol"
        case clientID = "client_id"
        case realityPublicKey = "reality_public_key"
        case shortID = "short_id"
        case serverName = "server_name"
        case flow
        case exitMode = "exit_mode"
        case maxSessions = "max_sessions"
        case maxMbps = "max_mbps"
        case relayVersion = "volunteer_version"
        case nodeClass = "node_class"
        case transport
        case wssFronts = "wss_fronts"
        case registeredAt = "registered_at"
        case lastHeartbeatAt = "last_heartbeat_at"
        case expiresAt = "expires_at"
        case city
        case country
        case countryCode = "country_code"
        case latitude
        case longitude
    }

    /// Explicit tolerant decoder: Swift's synthesized Decodable does not apply initializer defaults
    /// when a non-optional key is absent. Defaults here keep legacy signed descriptors direct-only.
    public init(from decoder: Decoder) throws {
        let values = try decoder.container(keyedBy: CodingKeys.self)
        id = try values.decode(String.self, forKey: .id)
        label = try values.decodeIfPresent(String.self, forKey: .label)
        publicHost = try values.decode(String.self, forKey: .publicHost)
        publicPort = try values.decode(Int.self, forKey: .publicPort)
        relayProtocol = try values.decode(String.self, forKey: .relayProtocol)
        clientID = try values.decode(String.self, forKey: .clientID)
        realityPublicKey = try values.decode(String.self, forKey: .realityPublicKey)
        shortID = try values.decode(String.self, forKey: .shortID)
        serverName = try values.decode(String.self, forKey: .serverName)
        flow = try values.decode(String.self, forKey: .flow)
        exitMode = try values.decode(String.self, forKey: .exitMode)
        maxSessions = try values.decode(Int.self, forKey: .maxSessions)
        maxMbps = try values.decode(Int.self, forKey: .maxMbps)
        relayVersion = try values.decode(String.self, forKey: .relayVersion)
        nodeClass = try values.decodeIfPresent(String.self, forKey: .nodeClass)
            ?? RelayConstants.nodeClassVolunteer
        transport = try values.decodeIfPresent(String.self, forKey: .transport) ?? ""
        wssFronts = try values.decodeIfPresent([WssFrontDescriptor].self, forKey: .wssFronts) ?? []
        registeredAt = try values.decode(Date.self, forKey: .registeredAt)
        lastHeartbeatAt = try values.decode(Date.self, forKey: .lastHeartbeatAt)
        expiresAt = try values.decode(Date.self, forKey: .expiresAt)
        city = try values.decodeIfPresent(String.self, forKey: .city)
        country = try values.decodeIfPresent(String.self, forKey: .country)
        countryCode = try values.decodeIfPresent(String.self, forKey: .countryCode)
        latitude = try values.decodeIfPresent(Double.self, forKey: .latitude)
        longitude = try values.decodeIfPresent(Double.self, forKey: .longitude)
    }
}

public struct RelayListResponse: Decodable, Equatable, Sendable {
    public let count: Int
    public let serverTime: Date
    public let relays: [RelayDescriptor]

    enum CodingKeys: String, CodingKey {
        case count
        case serverTime = "server_time"
        case relays
    }
}

public extension RelayDescriptor {
    func isUsable(now: Date = Date()) -> Bool {
        relayProtocol == RelayConstants.protocolVLESSRealityVision &&
            flow == RelayConstants.flowVision &&
            exitMode == RelayConstants.exitModeDirect &&
            expiresAt > now &&
            publicHost.isEmpty == false &&
            publicPort > 0 &&
            clientID.isEmpty == false &&
            realityPublicKey.isEmpty == false &&
            shortID.isEmpty == false &&
            serverName.isEmpty == false
    }

    /// Human-readable exit location such as "Tokyo, Japan", or "" while the broker has no geo.
    func locationLabel() -> String {
        [city, country].compactMap { $0 }.filter { $0.isEmpty == false }.joined(separator: ", ")
    }
}

File: android/app/src/main/java/com/openrung/model/RelayDescriptor.kt
package com.openrung.model

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable
import kotlinx.serialization.KSerializer
import kotlinx.serialization.SerializationException
import kotlinx.serialization.descriptors.SerialDescriptor
import kotlinx.serialization.descriptors.buildClassSerialDescriptor
import kotlinx.serialization.descriptors.element
import kotlinx.serialization.encoding.Decoder
import kotlinx.serialization.encoding.Encoder
import kotlinx.serialization.json.JsonDecoder
import kotlinx.serialization.json.JsonEncoder
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.intOrNull
import kotlinx.serialization.json.put
import java.time.Instant

object RelayConstants {
    const val PROTOCOL_VLESS_REALITY_VISION = "vless-reality-vision"
    const val FLOW_VISION = "xtls-rprx-vision"
    const val EXIT_MODE_DIRECT = "direct"
    const val TRANSPORT_DIRECT = "direct"
    const val NODE_CLASS_FOUNDATION = "foundation"
    const val NODE_CLASS_VOLUNTEER = "volunteer"
}

/** A signed WSS/CDN front advertised by the relay directory. */
@Serializable(with = WssFrontDescriptorSerializer::class)
data class WssFrontDescriptor(
    val id: String,
    val url: String,
    @SerialName("protocol_version")
    val protocolVersion: Int,
)

/**
 * Keeps top-level relay decoding forward-compatible while making the security-sensitive signed
 * front shape exact. Otherwise `ignoreUnknownKeys` would discard a front field before wsscore's
 * strict validator could see and reject it.
 */
object WssFrontDescriptorSerializer : KSerializer<WssFrontDescriptor> {
    override val descriptor: SerialDescriptor = buildClassSerialDescriptor(
        "com.openrung.model.WssFrontDescriptor",
    ) {
        element<String>("id")
        element<String>("url")
        element<Int>("protocol_version")
    }

    override fun serialize(encoder: Encoder, value: WssFrontDescriptor) {
        val jsonEncoder = encoder as? JsonEncoder
            ?: throw SerializationException("WSS fronts require JSON encoding")
        jsonEncoder.encodeJsonElement(
            buildJsonObject {
                put("id", value.id)
                put("url", value.url)
                put("protocol_version", value.protocolVersion)
            },
        )
    }

    override fun deserialize(decoder: Decoder): WssFrontDescriptor {
        val jsonDecoder = decoder as? JsonDecoder
            ?: throw SerializationException("WSS fronts require JSON decoding")
        val value = jsonDecoder.decodeJsonElement() as? JsonObject
            ?: throw SerializationException("WSS front must be an object")
        val unknown = value.keys - ALLOWED_KEYS
        if (unknown.isNotEmpty()) {
            throw SerializationException("WSS front contains unknown fields: ${unknown.sorted()}")
        }

        fun requiredString(name: String): String {
            val primitive = value[name] as? JsonPrimitive
            if (primitive == null || !primitive.isString) {
                throw SerializationException("WSS front $name must be a string")
            }
            return primitive.content
        }

        val protocol = value["protocol_version"] as? JsonPrimitive
        if (protocol == null || protocol.isString || protocol.intOrNull == null) {
            throw SerializationException("WSS front protocol_version must be an integer")
        }
        return WssFrontDescriptor(
            id = requiredString("id"),
            url = requiredString("url"),
            protocolVersion = checkNotNull(protocol.intOrNull),
        )
    }

    private val ALLOWED_KEYS = setOf("id", "url", "protocol_version")
}

@Serializable
data class RelayDescriptor(
    val id: String,
    /** Friendly relay name (operator-supplied or generated); empty on older brokers. */
    val label: String = "",
    @SerialName("public_host")
    val publicHost: String,
    @SerialName("public_port")
    val publicPort: Int,
    @SerialName("protocol")
    val relayProtocol: String,
    @SerialName("client_id")
    val clientId: String,
    @SerialName("reality_public_key")
    val realityPublicKey: String,
    @SerialName("short_id")
    val shortId: String,
    @SerialName("server_name")
    val serverName: String,
    val flow: String,
    @SerialName("exit_mode")
    val exitMode: String,
    @SerialName("max_sessions")
    val maxSessions: Int,
    @SerialName("max_mbps")
    val maxMbps: Int,
    /** Software version for any relay class; serialized name preserves the legacy broker wire field. */
    @SerialName("volunteer_version")
    val relayVersion: String,
    /** Trust class assigned by the signed directory; older descriptors are volunteer relays. */
    @SerialName("node_class")
    val nodeClass: String = RelayConstants.NODE_CLASS_VOLUNTEER,
    /** "direct" when clients reach this relay directly, "tunnel" when publicHost is a RelayHub. */
    val transport: String = "",
    /** Canonical, unique, sorted WSS/CDN fronts covered by the directory signature. */
    @SerialName("wss_fronts")
    val wssFronts: List<WssFrontDescriptor> = emptyList(),
    /** Whether the tunnel relay and its hub negotiated direct NAT punching. */
    @SerialName("punch_capable")
    val punchCapable: Boolean = false,
    /** HTTPS base URL of the hub's punch coordinator; supplied by the signed directory. */
    @SerialName("punch_endpoint")
    val punchEndpoint: String = "",
    @SerialName("registered_at")
    val registeredAt: String,
    @SerialName("last_heartbeat_at")
    val lastHeartbeatAt: String,
    @SerialName("expires_at")
    val expiresAt: String,
    // Broker-served exit location, absent until the broker's geo lookup succeeds (older brokers
    // never send it). For tunnel (CGNAT) relays this is where traffic actually exits, which is
    // NOT publicHost (the relay hub) — never geolocate publicHost client-side.
    val city: String = "",
    val country: String = "",
    @SerialName("country_code")
    val countryCode: String = "",
    val latitude: Double? = null,
    val longitude: Double? = null,
) {
    fun isUsable(now: Instant): Boolean {
        val expires = runCatching { Instant.parse(expiresAt) }.getOrNull() ?: return false
        return relayProtocol == RelayConstants.PROTOCOL_VLESS_REALITY_VISION &&
            flow == RelayConstants.FLOW_VISION &&
            exitMode == RelayConstants.EXIT_MODE_DIRECT &&
            expires > now &&
            publicHost.isNotBlank() &&
            publicPort > 0 &&
            clientId.isNotBlank() &&
            realityPublicKey.isNotBlank() &&
            shortId.isNotBlank() &&
            serverName.isNotBlank()
    }

    /** Human-readable exit location such as "Tokyo, Japan", or "" while the broker has no geo. */
    fun locationLabel(): String = listOf(city, country).filter { it.isNotBlank() }.joinToString(", ")

    /**
     * Relay name as surfaced in the UI (connect card status row, recents pills, log lines).
     * [label] is operator-supplied free text, so strip control/format characters (a bidi
     * override could reorder or spoof surrounding UI text), collapse whitespace, and clamp the
     * length, falling back to the broker-assigned id when nothing printable remains. Keep in
     * sync with the Swift `RelayDescriptor.displayName()` and the TS `sanitizeRelayName()`
     * in `src/model/exitNode.ts`.
     */
    fun displayName(): String = sanitizeDisplayName(label).ifEmpty { sanitizeDisplayName(id) }

    private companion object {
        const val MAX_DISPLAY_NAME_CODE_POINTS = 24

        fun sanitizeDisplayName(raw: String): String {
            // Iterate code points, not chars, so astral-plane format characters
            // (e.g. U+E00xx tags) are classified whole rather than as surrogates. After the
            // control/format strip the only whitespace left is the space separators
            // (Zs/Zl/Zp = Character.isSpaceChar) — collapse those; tab/newline are CONTROL
            // and already gone. Same rules as the Swift and TS sanitizers.
            val collapsed = buildString(raw.length) {
                var pendingSpace = false
                var i = 0
                while (i < raw.length) {
                    val codePoint = raw.codePointAt(i)
                    i += Character.charCount(codePoint)
                    when {
                        Character.getType(codePoint) == Character.CONTROL.toInt() ||
                            Character.getType(codePoint) == Character.FORMAT.toInt() -> Unit
                        Character.isSpaceChar(codePoint) -> pendingSpace = isNotEmpty()
                        else -> {
                            if (pendingSpace) {
                                append(' ')
                                pendingSpace = false
                            }
                            appendCodePoint(codePoint)
                        }
                    }
                }
            }
            // Clamp by code points (never mid-surrogate-pair by construction), matching
            // Swift's unicode-scalar prefix and TS's Array.from slice.
            var end = 0
            var count = 0
            while (end < collapsed.length && count < MAX_DISPLAY_NAME_CODE_POINTS) {
                end += Character.charCount(collapsed.codePointAt(end))
                count += 1
            }
            return collapsed.substring(0, end).trimEnd(' ')
        }
    }
}

@Serializable
data class RelayListResponse(
    val count: Int,
    @SerialName("server_time")
    val serverTime: String,
    val relays: List<RelayDescriptor>,
    // Relay-list signing metadata (SPEC v1 §2.2) remains part of the decoded wire model.
    // brokerapi verifies these fields before returning RelayJSON to Kotlin.
    /** RFC3339 expiry of this snapshot: `server_time` + 30 min on the API channel. */
    @SerialName("not_after")
    val notAfter: String = "",
    /** Advisory id (first 8 SHA-256 bytes of the signing pubkey, hex). */
    @SerialName("key_id")
    val keyId: String = "",
    /** "api" or "mirror"; already verified by brokerapi. */
    val channel: String = "",
    /** API channel: echo of the requested `limit`, already verified by brokerapi. */
    val limit: Int? = null,
) {
    val serverInstant: Instant
        get() = Instant.parse(serverTime)
}

@Serializable
data class ErrorResponse(
    val error: String = "",
)
